wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceSubjectRegistry.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.core.notification.EventSubscriber;
+import org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent;
+import org.eclipse.che.commons.env.EnvironmentContext;
+import org.eclipse.che.commons.subject.Subject;
+import org.slf4j.Logger;
+
+/**
+ * This class allows maintaining a link between a started workspace and the user who started it.
+ * This also allows updating, at any time, the user informations (<code>Subject</code>) of users
+ * that started workspaces.
+ *
+ * <p></br>
+ *
+ * <p>In particular, this allows having access to the up-to-date connection information (userName
+ * and Keycloak token) of the user that was used when creating a workspace.
+ *
+ * <p>This is required for all the use-cases where these user informations would be necessary to
+ * perform batch-like operations on the user workspaces (such as idling, stop at shutdown, etc ...).
+ *
+ * <p></br>
+ *
+ * <p>An example of such use-case is the multi-tenant scenario when deployment is done to Openshift:
+ * the user connection informations are required to have access to the Openshift cluster / namespace
+ * where the workspace has been created.
+ *
+ * @author David Festal
+ */
+@Singleton
+public class WorkspaceSubjectRegistry implements EventSubscriber<WorkspaceStatusEvent> {
+
+  private static final Logger LOG = getLogger(WorkspaceSubjectRegistry.class);
+
+  private final EventService eventService;
+  private final ReadWriteLock lock = new ReentrantReadWriteLock();
+  private final Map<String, Subject> workspaceOwners = new HashMap<>();
+  private final Multimap<String, String> userIdToWorkspaces =
```

Comment:
@davidfestal I propose to replace Multimap and Map to next mapping:

``` Java
Map<String, String> workspaceId2UserId
Map<String, Subject> userId2Subject
```
cases:

**1. when the workspace is stopped**
``` Java
...
String userId = workspaceId2UserId.remove(workspaceId);
if(userId != null) {
    userId2Subject.remove(userId);
}
...
```

**2. when the workspace is started**
``` Java
...
workspaceId2UserId.put(workspaceId, subject.getUserId());
updateSubject(Subject subject);
...

public void updateSubject(Subject subject) {
    String token = subject != null ? subject.getToken() : null;
    if (token != null && token.startsWith("machine")) {      
      return;
    }
    ...
    userId2Subject.put(subject.getUserId(), subject);
    ...
  }
```

**3. get subject by workpsace id:**

``` Java
public Subject getWorkspaceStarter(String workspaceId) {
  ...
  String userId = workspaceId2UserId.get(workspaceId);
  if(userId != null) {
    return userId2Subject.get(userId);     
  }
  return null;
  ...
}
```
WDYT?

Reply:
It seems to me that this proposal doesn't support starting 2 workspaces with the same user and then stopping only one of both. The user Subject would be lost though it is still useful for the second workspace.

Or did I miss something ?

Reply:
yes, I missed that and the first case when the workspace is stopped should be look like this:
``` Java
String userId = workspaceId2UserId.remove(workspaceId);
if(userId != null && !workspaceId2UserId.values().contains(userId)) {
    userId2Subject.remove(userId);
}
```

Reply:
also it seems to me that in this proposal, `updateSubject(Subject subject)` would allow adding in the `userId2Subject` map a `userId` that wasn't previously associated to a workspace. In such a case, it will never be removed from the map and we have a memory leak, no ?

Could you explain what seems like a problem for you in the current implementation ?

Reply:
Basically leak that you described is possible, but I thought that the start of the workspace it is the only case when this method is used and it can be solved by adding a check for mapping like:
`if(workspaceId2UserId.values().contains(userId)) {
     userId2Subject.put(subject.getUserId(), subject);
 }`

What I'm trying to improve it is the mapping of 
1 user to n workspaces and n workspaces to 1 subject
to
n workspaces to 1 user and 1 user to 1 subject.

Do you see what I mean?

Reply:
Yes, I see, but I'm not sure performance would be better in the `updateSubject()` method, knowing that you potentially need to iterate the whole list of started workspaces.

And the `updateSubject()` method should be as fast as possible, since it's called each time a client request comes to the wsmaster.

Another point: with your proposal, `getWorkspaceStarter()` requires 2 hash map `get` calls instead  of one.

In fact the idea of the current implementation was to optimize performance in those 2 methods that are called much more often than the workspace start or stop methods.

Reply:
@davidfestal I see, so let it be as it is.